File: 2021/Book/again/04-2. 게임개발.js
/**
 * 1. 현재 방향 기준, 왼쪽부터 차례대로 갈 곳을 정한다.
 * 2. 왼쪽에 가보지 않은 칸이 있으면, 그리로 전진. 없으면 다시 회전
 * 3. 네 방향이 모두 가본 칸이거나 바다로 되어있는 경우, 방향 유지한 채 한칸 뒤로가고 1번 진행.
 * 4. 뒤가 바다라면 종료
 */
const info = [1, 1, 0];
const map = [
  [1, 1, 1, 1],
  [1, 0, 0, 1],
  [1, 1, 0, 1],
  [1, 1, 1, 1],
];

function solution(info, map) {
  let [currentX, currentY, direction] = info;
  const mlen = map.length;
  // 북 동 남 서 --- direction 기준으로
  const dx = [-1, 0, 1, 0];
  const dy = [0, 1, 0, -1];
  let count = 0;
  // 플레이어 위치 방문 처리
  let land = 1;
  map[currentX][currentY] = 2;
  while (true) {
    direction = turnLeft(direction);
    // 다음 방향
    let nextX = currentX + dx[direction];
    let nextY = currentY + dy[direction];

    // 다음 방향이 0이면 전진
    if (map[nextX][nextY] === 0) {
      count = 0;
      land++;
      map[nextX][nextY] = 2;
      currentX = nextX;
      currentY = nextY;
      continue;
    } else count += 1;

    // count가 4이면 한 바퀴 돈 것이므로 갈 수 있는 방향이 없는 것. 뒤로 후진
    if (count === 4) {
      nextX = currentX - dx[direction];
      nextY = currentY - dy[direction];
      if (map[nextX][nextY] === 1) break;
      else {
        currentX = nextX;
        currentY = nextY;
        count = 0;
      }
    }
  }
  return land;
}

function turnLeft(direction) {
  direction--;
  if (direction === -1) direction = 3;
  return direction;
}

console.log(solution(info, map));

File: 2021/Book/again/04-2. 게임개발.py
# 맵 N x M
# 움직임
# 1. 왼쪽으로 회전
# 2. 회전한 뒤 앞 칸이 안가본 곳이면 전진, 아니면 왼쪽으로 다시 회전
# 3. 네 방향 모두 갈 곳이 없다면 뒤로 후진, 만약 뒤가 바다면 break

# 맵 크기
N, M = map(int, input().split())
# 캐릭터 위치 및 방향
x, y, d = map(int, input().split())
# 맵 설정
maps = []
for _ in range(N) :
    row = list(map(int, input().split()))
    maps.append(row)

# 방향 설정
directions = [0, 1, 2, 3]
dx = [-1, 0, 1, 0]
dy = [0, 1, 0, -1]

# 좌측으로 90도 회전
def turnLeft() :
    # 전역변수 사용
    global d, turn
    d -= 1
    # 방향이 범위를 넘을 시
    if d == -1 : d = 3
    # 회전횟수 증가
    turn += 1

# 주인공 현재 위치 표시
maps[x][y] = 2
turn = 0
count = 1
# 탐색 시작
while True :
    turnLeft()
    nextX = x + dx[d]
    nextY = y + dy[d]

    # 전진
    if maps[nextX][nextY] == 0 :
        x = nextX
        y = nextY
        maps[nextX][nextY] = 2
        turn = 0
        count += 1
        continue

    # 후진
    if turn == 4 :
        nextX = x - dx[d]
        nextY = y - dx[d]
        # 바다일 때
        if maps[nextX][nextY] == 1 :
            break
        else :
            x = nextX
            y = nextY
            turn = 0

print(count)